Base handler test

boekhorstb1 opened the pull request:
Added tests:
- checking for types
- checking for expected erros
- checking that functions pass variables
- checking that functions just run (no asserts)

boekhorstb1 committed:
Merge pull request #3 from maintaina-com/FRAMEWORK_6_0

merging from maintaina-com

boekhorstb1 committed:
setting phpunit.xml to have colors=true

boekhorstb1 committed:
initial commit

boekhorstb1 committed:
first test: checking that abstract function write must return a boolean

boekhorstb1 committed:
added tests for SetOption method

boekhorstb1 committed:
added last filterTest without assertions

boekhorstb1 committed:
made more explicit how types are tested

boekhorstb1 committed:
changed function to use log instead of write and explained in added comment

boekhorstb1 committed:
adding NullHandlerTest

boekhorstb1 committed:
adapted naming conventions

boekhorstb1 committed:
added another test of the write() method by implementing the MockHandler

boekhorstb1 commented:
I added a test for the Nullhandler and a test for the Mockhandler (both testing the write() method).

I noticed the errors in phpstan: they appearded to be because of a lack of backward comatibility. So I checked in my php7 environment, but got no errors

boekhorstb1 committed:
Merge pull request #5 from maintaina-com/FRAMEWORK_6_0

pulling to get latest updates form maintana-com/log

Commit:
automatic generation of composer.json - [ci skip]

midahp commented on test/Handler/BaseHandlerTest.php:
```diff
@@ -0,0 +1,89 @@
+    /**
+     * @doesNotPerformAssertions
+     */
+    public function testFiltersAreAdded(): void
```
This does not really test if the filters are added right? It only tests that the method exists and does not fail.

boekhorstb1 replied:
Ok will try to add better tests for adding filters. I fear that I cannot create a mock-object of LogMessage  easily (error: formatted message must not be accessed before initialisation), so will take the changed Mockhandler from Logger (is that ok?)

midahp reviewed (commented):
Looks good.

Can you also test whether the  `log` method uses the `filters` ?
You could `expect` the `formatMessage` method on the event to be called if it passes the filters.

boekhorstb1 committed:
added stuff to Mockhandler and made save for current functions in BasehandlerTest

boekhorstb1 committed:
Merge branch 'maintaina-com:FRAMEWORK_6_0' into FRAMEWORK_6_0

boekhorstb1 committed:
merged with latest updates from FRAMWORK 6_0 branch

boekhorstb1 committed:
took updated Mockhandler from Logger and added tests

boekhorstb1 committed:
giving type to Mockhandler check variable

midahp commented on test/Handler/BaseHandlerTest.php:
```diff
@@ -0,0 +1,120 @@
+    public function setUp(): void
+    {
+        # Bult in Mock for abstract classes (in phpunit)
+        $this->stub = $this->getMockForAbstractClass(BaseHandler::class);
```
Please use a different name like `baseHandlerMock` for this.

boekhorstb1 replied:
ok

midahp commented on test/Handler/BaseHandlerTest.php:
```diff
@@ -0,0 +1,120 @@
+        $this->message1 = 'this is an emergency!';
+        $this->logMessage1 = new LogMessage($this->level1, $this->message1, ['randomfield' => 'stuff']);
+
+        $this->constraint_filter = new ConstraintFilter();
```
We ususally use camelCase for methods, attributes and variables.

boekhorstb1 replied:
ok, will keep in mind

midahp commented on test/Handler/BaseHandlerTest.php:
```diff
@@ -0,0 +1,120 @@
+     */
+    public function testAbstractWriteFunctionsMustReturnBoolWithPHPunitAbstractClass(): void
+    {
+        $this->expectException('TypeError');
```
Expecting this exception seems more like a workaround than a test.
Use `$stub->method('write')->willReturn(true);`

boekhorstb1 replied:
mmm. So, when I do this:

````
 public function testAbstractWriteFunctionsMustReturnBoolWithPHPunitAbstractClass(): void
    {

        $baseHandlerMock = $this->baseHandlerMock;


         $baseHandlerMock->method('write')->willReturn(true);


        $this->assertTrue($baseHandlerMock->log($this->logMessage1));
    }
````

I get this error: 

````
1) Horde\Log\Test\Handler\BaseHandlerTest::testAbstractWriteFunctionsMustReturnBoolWithPHPunitAbstractClass
Failed asserting that null is true.
````

Maybe because the write() method is implemented from LogHandler? Maybe because BaseHandler  is an abstract class?

Will see if there is another way to test this or if I overlooked something

boekhorstb1 replied:
changed testMethod and its name to testWriteFunctionIsExecutedByLog()

midahp commented on test/Handler/BaseHandlerTest.php:
```diff
@@ -0,0 +1,120 @@
+
+        $stub->expects($this->once())
+                 ->method('write')
+                 ->will($this->returnValue(null));
```
At least for me this did not work. `->willReturn(true)` worked for me

midahp commented on test/Handler/BaseHandlerTest.php:
```diff
@@ -0,0 +1,120 @@
+
+    public function testAbstractWriteFunctionsWithOwnMockClass()
+    {
+        $this->assertTrue($this->mockhandler->write($this->logMessage1));
```
You are not testing the BaseHandler here?

boekhorstb1 replied:
````
class MockHandler extends BaseHandler
````

The problem with Basehandler was that it is abstract, I think. So in order to avoid trouble I implemtented it in a 'normal' class, the Mockclass. (I think it was made for that purpose)

midahp commented on test/Handler/BaseHandlerTest.php:
```diff
@@ -0,0 +1,120 @@
+    {
+        $this->expectException(LogException::class);
+
+        $this->mockhandler->setOption('foo', 'bar');
```
This should be done in a separate testfile for the SetOptionsTrait.

boekhorstb1 replied:
ok created seperate Testfile

midahp commented on test/Handler/BaseHandlerTest.php:
```diff
@@ -0,0 +1,120 @@
+
+    public function testSetOptionReturnsTrueWhenCorrectParams(): void
+    {
+        $this->assertTrue($this->mockhandler->setOption('ident', 'test'));
```
This should be done in a separate testfile for the SetOptionsTrait.

boekhorstb1 replied:
a ok can do that. I thought I should test setOption because it is part of the LogHandler which is implemetned by the BaseHandler

midahp reviewed (changes requested):
Please see the comments.

boekhorstb1 committed:
renamed variable stub to basehandlermock

boekhorstb1 committed:
replaced constraint_filter with constraintFilter

boekhorstb1 committed:
testing that the log function is executed

boekhorstb1 committed:
rewrote testWriteFunctionReturnsBoolean()

boekhorstb1 committed:
refactored code: Basehandlertest and SetOptionTraitTest

boekhorstb1 committed:
small cleanup changes

boekhorstb1 commented:
Ok I did the following:
- renamed variables to camelCase 
- renamed methods  and mocks more appropriatly.
- changed the testmethods concerning the write() method
- create a seperate file for testint SetOptionsTrait

boekhorstb1 committed:
renamed Filtertest

midahp commented on test/Handler/SetOptionTraitTest.php:
```diff
@@ -0,0 +1,40 @@
+{
+    public function setUp(): void
+    {
+        $this->mockhandler = new MockHandler();
```
Try to not use the mockhandler here. Try this instead:
```
$this->setOptionsTrait = $this->getMockForTrait(SetOptionsTrait::class);
$this->setOptionsTrait->options = new Options();
```

boekhorstb1 replied:
ok removed mockhandler from the test

boekhorstb1 committed:
adding tests for SetOptionTrait wihtout Mockhandler

boekhorstb1 committed:
remvoing mockhandler tests
